autofaiss/external/distributed.py

```diff
@@ -0,0 +1,349 @@
+    if ss is None:
+        print("No pyspark session found, creating a new one!")
+        ss = (
+            pyspark.sql.SparkSession.builder.config("spark.driver.memory", "16G")
+            .master("local[1]")
+            .appName("Distributed autofaiss")
+            .getOrCreate()
+        )
+    return ss
+
+
+def _batch_loader(batch_size: int, nb_batches: int) -> Iterator[Tuple[int, int, int]]:
+    """Yield [batch id, batch start position, batch end position]"""
+    for batch_id in range(nb_batches):
+        start = batch_size * batch_id
+        end = batch_size * (batch_id + 1)
+        yield batch_id, start, end
+
+
+def _download_one(src_dst_path: Tuple[str, str]):
+    src_path, dst_path = src_dst_path
+    os.system(f"hdfs dfs -copyToLocal {src_path} {dst_path}")
+
+
+def _parallel_download_indices_from_hdfs(indices_file_paths: List[str], dst_folder: str):
+    """Download small indices in parallel."""
+    if len(indices_file_paths) == 0:
+        return
+    os.makedirs(dst_folder, exist_ok=True)
+    nb_files = len(indices_file_paths)
+    src_paths = ("hdfs://root" + filepath for filepath in indices_file_paths)
+    src_dest_paths = zip(src_paths, repeat(dst_folder))
+    with tqdm(total=nb_files) as pbar:
+        with multiprocessing.Pool(processes=min(8, multiprocessing.cpu_count())) as pool:
+            for _ in pool.imap_unordered(_download_one, src_dest_paths):
+                pbar.update(1)
+
+
+def _merge_index(small_indices_folder: str) -> faiss.Index:
+    """Merge all the indices in `small_indices_folder` into single one return the merged index."""
+    fs = _get_file_system(small_indices_folder)
+    small_indices_files = fs.ls(small_indices_folder, detail=False)
+    if len(small_indices_files) == 0:
+        raise ValueError(f"No small index is saved in {small_indices_folder}")
+
+    def _get_index_from_file(filepath: str) -> faiss.Index:
+        with open(filepath, "rb") as f:
+            index_bytes = f.read()
+        return _get_index_from_bytes(index_bytes)
+
+    def _merge_from_local() -> faiss.Index:
+        local_file_paths = [
+            os.path.join(local_indices_folder, filename) for filename in os.listdir(local_indices_folder)
+        ]
+        with Timeit("-> Load first index", indent=4):
+            merged = _get_index_from_file(local_file_paths[0])
+        with Timeit("-> Merge the rest of indices", indent=4):
+            for rest_index_file in tqdm(local_file_paths[1:]):
+                index = _get_index_from_file(rest_index_file)
+                faiss.merge_into(merged, index, shift_ids=True)
+        return merged
+
+    local_indices_folder = "/tmp/distributed_autofaiss_indices"
+    with Timeit("-> Download small indices from HDFS to local", indent=4):
+        _parallel_download_indices_from_hdfs(indices_file_paths=small_indices_files, dst_folder=local_indices_folder)
```

Comment:
this will fail if you don't have enough room to download 100% of indices at once
It will be better to first estimate the index sizes, then download N at a time such that N*size_index <= max_size_on_disk with max_size_on_disk default to 100GB